src/OrleansRuntime/GrainDirectory/GrainDirectoryPartition.cs

```diff
@@ -364,21 +376,58 @@ internal void RemoveGrain(GrainId grain)
         internal AddressesAndTag LookUpActivations(GrainId grain)
```

Comment:
Since the overwhelming majority of cases, pretty close to100%, involves a single activation per grain registered in the directory, wouldn't it make sense to optimize that hot path instead of the general but non-existent in reality case of multiple activations registered per grain? For example, would it make sense to add a field to `AddressesAndTag` of type `ActivationAddress` for a single result, so that the `Addresses` fields will pretty much always be set to null and we'd never allocate those lists? Alternatively, we could make lookup methods return a `Tuple<AddressesAndTag, AddressAndTag>`. Although a Tuple would cause another undesirable allocation, so maybe adding a field would be cheaper. C# 7 will eventually allow to return multiple values from a method more naturally.